godot-core/src/obj/oneditor.rs

```diff
@@ -0,0 +1,293 @@
+///     let mut my_node = SomeClassThatCanBeInstantiatedInCode::new_alloc();
+///
+///     // Would cause the panic:
+///     // this.add_child(&my_node);
+///
+///     my_node.bind_mut().some_primitive.init(val);
+///
+///     // Will not cause the panic.
+///     this.add_child(&my_node);
+///
+///     my_node
+/// }
+/// ```
+///
+pub struct OnEditor<T> {
+    inner: OnEditorState<T>,
+}
+
+enum OnEditorState<T> {
+    /// Uninitialized null value.
+    Null,
+    /// Uninitialized state, but with a value marked as invalid (required to represent non-nullable type in the editor).
+    Uninitialized(T),
+    /// Initialized with a value.
+    Initialized(T),
+}
+
+impl<T: GodotConvert + Var + FromGodot + PartialEq> OnEditor<T> {
+    /// Initializes `OnEditor<T>` with given value.
+    ///
+    /// Panics if `init()` was called before.
+    pub fn init(&mut self, val: T) {
+        match self.inner {
+            OnEditorState::Null | OnEditorState::Uninitialized(_) => {
+                *self = OnEditor {
+                    inner: OnEditorState::Initialized(val),
+                };
+            }
+            OnEditorState::Initialized(_) => {
+                panic!("godot-rust: Given OnEditor value has been already initialized; did you call init() more than once?")
+            }
+        }
+    }
+
+    /// Creates new uninitialized `OnEditor<T>` with value which will be used to represent invalid, uninitialized property in the editor.
+    pub fn uninit(val: T) -> Self
+    where
+        T: DirectExport,
+    {
+        OnEditor {
+            inner: OnEditorState::Uninitialized(val),
+        }
+    }
+
+    /// Creates new uninitialized `OnEditor<T>` value for nullable GodotTypes.
+    ///
+    /// Not a part of public API – available only via `Default` implementation on `OnEditor<Gd<T>>` and `OnEditor<DynGd<D, T>>`.
+    pub(crate) fn null() -> Self {
+        OnEditor {
+            inner: OnEditorState::Null,
+        }
+    }
+
+    #[doc(hidden)]
+    pub fn is_invalid(&self) -> bool {
+        match self.inner {
+            OnEditorState::Null | OnEditorState::Uninitialized(_) => true,
+            OnEditorState::Initialized(_) => false,
+        }
+    }
+
+    /// `Var::get_property` implementation that works both for nullable and non-nullable types.
+    pub(crate) fn get_property_inner(&self) -> Option<T::Via> {
+        match &self.inner {
+            OnEditorState::Null => None,
+            OnEditorState::Uninitialized(val) | OnEditorState::Initialized(val) => {
+                Some(val.get_property())
+            }
+        }
+    }
+
+    /// `Var::set_property` implementation that works both for nullable and non-nullable types.
+    pub(crate) fn set_property_inner(&mut self, value: Option<T::Via>)
+    where
+        T::Via: PartialEq,
+    {
+        match (value, &mut self.inner) {
+            (None, _) => self.inner = OnEditorState::Null,
```

Comment:
Is it valid to reset the state to null at any time? Independent of its current state?

Reply:
Thank you!